pkg/generic/thrift/http.go

```diff
@@ -77,12 +108,83 @@ func (r *ReadHTTPResponse) SetBase64Binary(enable bool) {
 	r.base64Binary = enable
 }
 
+// SetUseRawBodyForHTTPResp ...
+func (r *ReadHTTPResponse) SetUseRawBodyForHTTPResp(useRawBodyForHTTPResp bool) {
+	r.useRawBodyForHTTPResp = useRawBodyForHTTPResp
+}
+
+// SetDynamicGo ...
+func (r *ReadHTTPResponse) SetDynamicGo(cv t2j.BinaryConv, msg remote.Message) {
+	r.cv = cv
+	r.msg = msg
+	r.dynamicgoEnabled = true
+}
+
 // Read ...
 func (r *ReadHTTPResponse) Read(ctx context.Context, method string, in thrift.TProtocol) (interface{}, error) {
-	fnDsc, err := r.svc.LookupFunctionByMethod(method)
+	// fallback logic
+	if !r.dynamicgoEnabled {
+		fnDsc, err := r.svc.LookupFunctionByMethod(method)
+		if err != nil {
+			return nil, err
+		}
+		fDsc := fnDsc.Response
+		resp, err := skipStructReader(ctx, in, fDsc, &readerOption{forJSON: true, http: true, binaryWithBase64: r.base64Binary})
+		if r.useRawBodyForHTTPResp {
+			if httpResp, ok := resp.(*descriptor.HTTPResponse); ok && httpResp.Body != nil {
+				rawBody, err := customJson.Marshal(httpResp.Body)
+				if err != nil {
+					return nil, err
+				}
+				httpResp.RawBody = rawBody
+			}
+		}
+		return resp, err
+	}
+
+	// dynamicgo logic
+	if r.msg.PayloadLen() == 0 {
+		return nil, perrors.NewProtocolErrorWithMsg("msg.PayloadLen should always be greater than zero")
+	}
+
+	tProt, ok := in.(*cthrift.BinaryProtocol)
+	if !ok {
+		return nil, perrors.NewProtocolErrorWithMsg("TProtocol should be BinaryProtocol")
+	}
+
+	tyDsc := r.svc.DynamicGoDsc.Functions()[method].Response()
+
+	mBeginLen := bthrift.Binary.MessageBeginLength(method, thrift.TMessageType(r.msg.MessageType()), r.msg.RPCInfo().Invocation().SeqID())
+	sName, err := in.ReadStructBegin()
+	if err != nil {
+		return nil, err
+	}
+	sBeginLen := bthrift.Binary.StructBeginLength(sName)
+	fName, typeId, id, err := in.ReadFieldBegin()
 	if err != nil {
 		return nil, err
 	}
-	fDsc := fnDsc.Response
-	return skipStructReader(ctx, in, fDsc, &readerOption{forJSON: true, http: true, binaryWithBase64: r.base64Binary})
+	fBeginLen := bthrift.Binary.FieldBeginLength(fName, typeId, id)
+	transBuf, err := tProt.ByteBuffer().ReadBinary(r.msg.PayloadLen() - mBeginLen - sBeginLen - fBeginLen - bthrift.Binary.MessageEndLength())
+	if err != nil {
+		return nil, err
+	}
+	fid := dthrift.FieldID(id)
+
+	buf := make([]byte, 0, len(transBuf)*2)
```

Comment:
why *2 here?

Reply:
@felix021 This is because json size is usually 2x times larger than equivalent thrift data according to @AsterDY 's and my experiments. I'll add a comment for this line.

Reply:
and why we're not using `mcache.Malloc` here? It's used in the Writer

Reply:
fixed